packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -350,261 +352,236 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
       .contains(endOffset + effectiveOffset);
   }
 
-  static const int _kLeftArrowCode = 21;
-  static const int _kRightArrowCode = 22;
-  static const int _kUpArrowCode = 19;
-  static const int _kDownArrowCode = 20;
-  static const int _kXKeyCode = 52;
-  static const int _kCKeyCode = 31;
-  static const int _kVKeyCode = 50;
-  static const int _kAKeyCode = 29;
-  static const int _kDelKeyCode = 112;
-
-  // The extent offset of the current selection
-  int _extentOffset = -1;
-
-  // The base offset of the current selection
-  int _baseOffset = -1;
-
-  // Holds the last location the user selected in the case that he selects all
-  // the way to the end or beginning of the field.
-  int _previousCursorLocation = -1;
+  // Holds the last cursor location the user selected in the case the user tries
+  // to select vertically past the end or beginning of the field. If they do,
+  // then we need to keep the old cursor location so that we can go back to it
+  // if they change their minds. Only used for moving selection up and down in a
+  // multi-line text field when selecting using the keyboard.
+  int _cursorResetLocation = -1;
 
   // Whether we should reset the location of the cursor in the case the user
-  // selects all the way to the end or the beginning of a field.
-  bool _resetCursor = false;
-
-  static const int _kShiftMask = 1; // https://developer.android.com/reference/android/view/KeyEvent.html#META_SHIFT_ON
-  static const int _kControlMask = 1 << 12; // https://developer.android.com/reference/android/view/KeyEvent.html#META_CTRL_ON
-
-  // Call through to onSelectionChanged only if the given nextSelection is
-  // different from the existing selection.
-  void _handlePotentialSelectionChange(
+  // tries to select vertically past the end or beginning of the field. If they
+  // do, then we need to keep the old cursor location so that we can go back to
+  // it if they change their minds. Only used for resetting selection up and
+  // down in a multi-line text field when selecting using the keyboard.
+  bool _wasSelectingVerticallyWithKeyboard = false;
+
+  // Call through to onSelectionChanged.
+  void _handleSelectionChange(
     TextSelection nextSelection,
     SelectionChangedCause cause,
   ) {
-    if (nextSelection == selection) {
+    // Changes made by the keyboard can sometimes be "out of band" for listening
+    // components, so always send those events, even if we didn't think it
+    // changed.
+    if (nextSelection == selection && cause != SelectionChangedCause.keyboard) {
       return;
     }
     onSelectionChanged(nextSelection, this, cause);
   }
 
-  // TODO(goderbauer): doesn't handle extended grapheme clusters with more than one Unicode scalar value (https://github.com/flutter/flutter/issues/13404).
```

Comment:
Actually, you're right, this does still apply, but only because some parts of this code depend upon the length of a string in code points instead of extended grapheme clusters.  I'll restore it and add more info.